### PR TITLE
Add pytest - step 1

### PR DESCRIPTION
**Pull request by lekcyjna123**
Hi,
here is a PR which is a first step towards using pytest as described in #405. In this PR:
- ~a basic configuration for pytest is added~
- ~a new dependecy to requirements is added~
- updated some stubs
- removed pytest warnings:
  - removed deprecated `Repl`
  - ported deprecated `RoundRobin` from amaranth to transactron, according to the recomendation from amaranth
  - updated some old tests which break test naming convention, to use newer way of writing tests by using `SimpleTestCircuit`
  - ~added exception for `TestbenchIO` naming which breaks test naming convention~
  - ~added exception for deprecation warnings related with `Record`~

There is a one warning left, but it should be fixed under separate PR, because it hides some error in `test_retirement.py` and it has to be investigated deeper.

### CODE REVIEW

**Comment by tilk** [comment]
This is mostly a refactor PR. I believe the first two points related to pytest can be delayed for another PR.

Question: why does the testing framework even care about `TestbenchIO`? Could you elaborate?

**Comment by lekcyjna123** [comment]
`pytest` do an inspection over all modules which starts with `test_` or ends with `_test` and looks for the classes which starts from `Test` and try to interpret them as a test suites (see: https://docs.pytest.org/en/7.4.x/explanation/goodpractices.html#test-discovery).  We import `TestbenchIO` to nearly every test module and this class name starts with `Test`. So pytest tries to interpret `TestbenchIO` as a test suite which is not successful.

**Comment by tilk** [comment]
Is it possible to just exclude the directories which shouldn't contain tests? The link you gave suggests it's possible.

File: test/frontend/test_decode.py [hunk 1]
```diff
@@ -1,47 +1,37 @@
...
 from coreblocks.params.configurations import test_core_config
 
 
-class TestElaboratable(Elaboratable):
-    def __init__(self, gen_params: GenParams):
-        self.gen_params = gen_params
-
-    def elaborate(self, platform):
-        m = Module()
+class TestFetch(TestCaseWithSimulator):
```
Comment (piotro888) [inline test/frontend/test_decode.py:10]:
```suggestion
class TestDecode(TestCaseWithSimulator):
```
:)

**Review by piotro888 — APPROVED** [review]
Nice test simplification, I think that #200 can be linked.

What is the problem with `test_retirement`?

**Comment by lekcyjna123** [comment]
> Is it possible to just exclude the directories which shouldn't contain tests? The link you gave suggests it's possible.

It is possible, but not useful in our case, because `TestbenchIO` is imported in our test file namespace (by `from test.common import TestbenchIO`) and it is found in regular test files.

> test/common/testbenchio.py:11: 12 warnings
  .../coreblocks/test/common/testbenchio.py:11: PytestCollectionWarning: cannot collect test class 'TestbenchIO' because it has a \_\_init\_\_ constructor (from: test/structs_common/test_exception.py)
    class TestbenchIO(Elaboratable):

**Comment by lekcyjna123** [comment]
> What is the problem with `test_retirement`?

We try to control signal value before start of simulation:
```py
    def test_rand(self):
        self.retc = RetirementTestCircuit(self.gen_params)

        yield from self.retc.mock_fetch_stall.enable()  # To be fixed
        with self.run_simulation(self.retc) as sim:
             ...
```

**Comment by tilk** [comment]
> We try to control signal value before start of simulation:

This code doesn't make any sense. We missed it in #523.

**Review by tilk — APPROVED** [review]

**Pull request merged**